### PR TITLE
Fix spacing of links in inline code.

### PR DESCRIPTION
**Pull request by steffahn**
Similar to #80733, but the focus is different. This PR eliminates all occurrences of pieced-together inline code blocks like [`Box`]`<`[`Option`]`<T>>` and replaces them with good-looking ones (using HTML-syntax), like <code>[Box]<[Option]\<T>></code>. As far as I can tell, I should’ve found all of these in the standard library (regex search with `` r"`\]`|`\[`" ``) \[except for in `core::convert` where I’ve noticed other things in the docs that I want to fix in a separate PR]. In particular, unlike #80733, I’ve added almost no new instance of inline code that’s broken up into multiple links (or some link and some link-free part). I also added tooltips (the stuff in quotes for the markdown link listings) in places that caught my eye, but that’s by no means systematic, just opportunistic.

[Box]: https://doc.rust-lang.org/std/boxed/struct.Box.html "Box"
[`Box`]: https://doc.rust-lang.org/std/boxed/struct.Box.html "Box"
[Option]: https://doc.rust-lang.org/std/option/enum.Option.html "Option"
[`Option`]: https://doc.rust-lang.org/std/option/enum.Option.html "Option"

Context: I got annoyed by repeatedly running into new misformatted inline code while reading the standard library docs. I know that once issue #83997 (and/or related ones) are resolved, these changes become somewhat obsolete, but I fail to notice much progress on that end right now.

r? @jyn514

### CODE REVIEW

File: library/alloc/src/fmt.rs [hunk 1]
```diff
@@ -505,23 +505,19 @@
...
 //!
-//! [`fmt::Result`]: Result
-//! [`Result`]: core::result::Result
-//! [`std::fmt::Error`]: Error
-//! [`write!`]: core::write
-//! [`write`]: core::write
-//! [`format!`]: crate::format
-//! [`to_string`]: crate::string::ToString
-//! [`writeln!`]: core::writeln
+//! [`fmt::Result`]: Result "fmt::Result"
```
Comment (jyn514) [inline library/alloc/src/fmt.rs:508]: By default this won't have a tooltip, right? Why did you add one?

**Review by jyn514 — COMMENTED** [review]
Looked at the first couple commits and they seem fine, but there's a lot of churn adding tooltips and I'm not really sure why they're useful to add - would like to hear more about that before spending more time on review.

File: library/alloc/src/fmt.rs (hunk 1 above)
Comment (steffahn) [reply library/alloc/src/fmt.rs:508]: In general, links generated by rustdoc do have a tooltip.

E.g. something like
```rs
/// Talking about [`Result`][std::fmt::Result] here!
pub mod m {}
```
will come with a tooltip

Screenshot_20210906_223756

Whereas (unfortunately) the following alternative
```rs
/// Talking about [`Result`] here!
/// 
/// [`Result`]: std::fmt::Result
pub mod m {}
```
creates almost the same documentation, just lacking the tooltip. That’s the general reason why I’ve converted many instances of
```markdown
[foo]: path
```
into
```markdown
[foo]: path "path"
```
and in some cases, I’ve also adjusted the tooltip slightly to make it more easy to understand. This includes the case here because `Result` is part of the prelude, a simple `"Result"` tooltip seems confusing to me, the `std::` prefix is something I’ve often avoided for tooltips OTOH (not least because the paths themselves also often are relative so some imported `std::foo` module).

<hr>

Also, as mentioned in the description of this PR, the addition of tooltips isn’t systematic and only covers places that I happened to come across. (Furthermore, the amount of tooltips added declines from earlier to later commits.)

**Comment by steffahn** [comment]
> I'm not really sure why they're useful to add

For me it’s mostly a question of consistency. I don’t like _not_ having a tooltip simply because of the decision to not put the link inline but separately in a links section in markdown. I don’t like reading docs where half the links give me tooltips and the other half don’t that’s also the reason why I had systematically included tooltips in #80733.

File: library/alloc/src/fmt.rs (hunk 1 above)
Comment (jyn514) [reply library/alloc/src/fmt.rs:508]: Ok. I think this would be better to fix systematically in rustdoc rather than adding it piecemeal to individual links. Could you open an issue and revert the tooltips you've added this PR?

Comment (steffahn) [reply library/alloc/src/fmt.rs:508]: "revert" as in “add another commit to revert them” or as a “rebase / edit / force push”. Because the former could easily happen in a later PR, too; these tooltips aren't particularly hard to find (and remove) with a regex search.

Either way, the revert would only remove those cases where the path and tooltip look _identical_, right? E.g. the `fmt::Result` should better stay IMO, because I don't like a `Result` tooltip.

Comment (steffahn) [reply library/alloc/src/fmt.rs:508]: Hmm, no answer. I’m assuming that an additional commit is okay, and that only the ones where tooltip and path are identical need to be removed. This is more easily accomplished by a systematic find-and-replace that by looking through all my commits here. This could, as noted, easily be a separate PR, done *after* the relevant change to rustdoc was made, but I’m including a commit here now to get this PR going.

Comment (jyn514) [reply library/alloc/src/fmt.rs:508]: Sorry for the delay, I haven't had much time for rust lately.

> Either way, the revert would only remove those cases where the path and tooltip look identical, right? E.g. the fmt::Result should better stay IMO, because I don't like a Result tooltip.

Well, ideally I wouldn't want any changes, because it makes them inconsistent and it will hopefully be done by rustdoc automatically eventually ... but if you feel strongly it seems fine to keep, it's just more work for me to review.

> I’m assuming that an additional commit is okay, and that only the ones where tooltip and path are identical need to be removed.

Yup, that's fine - I don't have strong opinions about git commits.

Comment (steffahn) [reply library/alloc/src/fmt.rs:508]: > I think showing the link used in the source is more clear in a way but I don't feel strongly.

The problem I see is that the link used in the source code depends on what the imports are, and the imports depends on what happens to be used in the source code. If e.g. I’m writing documentation mentioning, say, `mem::transmute`, then I’d use the path `men::transmute` (resulting in a `men::transmute` tooltip) if the module I’m in does `use core::mem` (or `use crate::mem`). But if `mem` is **only** used in docs, AFAIK I can’t just add an import for that because (I think) it’ll trigger an unused imports lint and fail CI. But it’s IMO better to *concistently* have it have the tooltip `mem::transmute` and not sometimes `core::mem::transmute` or and `crate::mem::transmute`, etc…, depending on the implementation details (i.e. imports) of the module your documentation comes from (in particular, I don’t think the `crate::` prefix is a good thing showing up in standard library tooltips).

File: library/core/src/option.rs [hunk 2]
```diff
@@ -111,16 +111,20 @@
...
-//! * [`as_deref`] converts from `&Option<T>` to `Option<&T::Target>`
-//! * [`as_deref_mut`] converts from `&mut Option<T>` to
-//!   `Option<&mut T::Target>`
-//! * [`as_pin_ref`] converts from [`Pin`]`<&Option<T>>` to
-//!   `Option<`[`Pin`]`<&T>>`
-//! * [`as_pin_mut`] converts from [`Pin`]`<&mut Option<T>>` to
-//!   `Option<`[`Pin`]`<&mut T>>`
-//!
+//! * [`as_ref`] converts from <code>[&][][Option]\<T></code> to <code>[Option]<[&]T></code>
+//! * [`as_mut`] converts from <code>[&mut] [Option]\<T></code> to <code>[Option]<[&mut] T></code>
```
Comment (jyn514) [inline library/core/src/option.rs:115]: I assume these are doing odd things to avoid linking `&` to Option - is there a reason you use `[&][]` for one and `[&mut] ` for the other?

File: library/core/src/option.rs [hunk 3]
```diff
@@ -111,16 +111,20 @@
...
+//!   <code>[Option]<[&]T::[Target]></code>
+//! * [`as_deref_mut`] converts from <code>[&mut] [Option]\<T></code> to
+//!   <code>[Option]<[&mut] T::[Target]></code>
+//! * [`as_pin_ref`] converts from <code>[Pin]<[&][][Option]\<T>></code> to
+//!   <code>[Option]<[Pin]<[&]T>></code>
+//! * [`as_pin_mut`] converts from <code>[Pin]<[&mut] [Option]\<T>></code> to
+//!   <code>[Option]<[Pin]<[&mut] T>></code>
+//!
+//! [&]: ../../std/primitive.reference.html "shared reference"
+//! [&mut]: ../../std/primitive.reference.html "mutable reference"
```
Comment (jyn514) [inline library/core/src/option.rs:126]:
```suggestion
//! [&]: reference "shared reference"
//! [&mut]: reference "mutable reference"
```

File: library/core/src/option.rs [hunk 4]
```diff
@@ -649,7 +653,9 @@ impl<T> Option<T> {
         }
     }
 
-    /// Converts from [`Pin`]`<&Option<T>>` to `Option<`[`Pin`]`<&T>>`.
+    /// Converts from <code>[Pin]<[&]Option\<T>></code> to <code>Option<[Pin]<[&]T>></code>.
+    ///
+    /// [&]: ../../std/primitive.reference.html "shared reference"
```
Comment (jyn514) [inline library/core/src/option.rs:658]:
```suggestion
    /// [&]: reference "shared reference"
```

File: library/core/src/option.rs [hunk 5]
```diff
@@ -658,7 +664,9 @@ impl<T> Option<T> {
         unsafe { Pin::get_ref(self).as_ref().map(|x| Pin::new_unchecked(x)) }
     }
 
-    /// Converts from [`Pin`]`<&mut Option<T>>` to `Option<`[`Pin`]`<&mut T>>`.
+    /// Converts from <code>[Pin]<[&mut] Option\<T>></code> to <code>Option<[Pin]<[&mut] T>></code>.
+    ///
+    /// [&mut]: ../../std/primitive.reference.html "mutable reference"
```
Comment (jyn514) [inline library/core/src/option.rs:669]:
```suggestion
    /// [&mut]: reference "mutable reference"
```

**Review by jyn514 — COMMENTED** [review]
I would prefer not to review PRs this big in the future - please wait until some of the rustdoc issues are fixed before making more PRs.

File: library/core/src/option.rs (hunk 2 above)
Comment (steffahn) [reply library/core/src/option.rs:115]: There’s no space after the `&`. The `[]` seems to be able to separate the two links without changing the fact that the `[&]: ../../std/primitive.reference.html "shared reference"` link is used. The space after `[&mut]` suffices to avoid the same problem there.

File: library/core/src/option.rs (hunk 3 above)
Comment (steffahn) [reply library/core/src/option.rs:126]: oh, that works too!? great!

File: library/core/src/option.rs (hunk 2 above)
Comment (jyn514) [reply library/core/src/option.rs:115]: Right, but why not use the space in both cases?

Comment (steffahn) [reply library/core/src/option.rs:115]: Because we don’t want to render `& Option<T>` but `&Option<T>`

Comment (steffahn) [reply library/core/src/option.rs:115]: It’s conventionally `&T` and `&mut T`, one has a space, the other doesn’t.

Comment (jyn514) [reply library/core/src/option.rs:115]: Got it, that makes sense :+1:

File: library/alloc/src/sync.rs [hunk 6]
```diff
@@ -476,7 +477,7 @@ impl<T> Arc<T> {
     /// assert_eq!(*zero, 0)
     /// ```
     ///
-    /// [zeroed]: ../../std/mem/union.MaybeUninit.html#method.zeroed
+    /// [zeroed]: mem::MaybeUninit::zeroed
```
[the remark this answers is not in the document]
Comment (steffahn) [reply library/alloc/src/sync.rs:480]: I have no idea :-)

**Comment by jyn514** [comment]
r=me with the linkchecker passing. Not sure if you have bors privileges or not, just in case -

@bors delegate=steffahn

Please use `r=jyn514`, not r+.

**Comment by jyn514** [comment]
oh - and please squash the commits before merging too if you don't mind :)

**Comment by steffahn** [comment]
> Not sure if you have bors privileges

I don’t :-)

> please squash the commits before merging

okay

**Comment by steffahn** [comment]
Squashed ([no changes](https://github.com/rust-lang/rust/compare/0cd383569d45ebf86a7f98e223ff6c2cfb779dee..67065fe93366cfdd9e88440b6f876c3dfeb1f937))

@bors r=jyn514

**Comment by steffahn** [comment]
This “regression” is about how fast rustc compiles some specific crate, right? (I have no idea where to find more information about those test cases.) This PR only changes (doc) comments in the standard library, so the regression is either sporadic or very weird.

If it's a sporadic thing, the benchmark of the next PR that gets merged by bors should perform better again; we'll see.

**Comment by rylev** [comment]
An update: thanks to [great investigation](https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/investigation.20.2389343) by @Mark-Simulacrum, we've identified a possible cause for this which should be fixed by https://github.com/rust-lang/rust/pull/89408
